### PR TITLE
Testing + code coverage for Elias review

### PR DESCRIPTION
**Pull request by holgerteichgraeber**
- Addresses #91 
- connection to codecov.io : https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl .
- cleanup of some unused functions
- replaced `@error` by `error(msg)`. The former is used for logging, whereas the later throws an Exception. We want the latter.

The branch `testing` is based off `documentation`. Before evaluating this PR, please first review #96 . 

In #91, there are several TODOs concerning which functions still need test coverage. Let's implement them after this PR.

### CODE REVIEW

**Comment by holgerteichgraeber** [comment]
> 
> # [Codecov](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97?src=pr&el=h1) Report
> 
> > ❗️ No coverage uploaded for pull request base (`documentation@d483c9e`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference#section-missing-base-commit).
> > The diff coverage is `50%`.
> 
> [Impacted file tree graph](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97?src=pr&el=tree)
> 
> ```diff
> @@               Coverage Diff                @@
> ##             documentation      #97   +/-   ##
> ================================================
>   Coverage                 ?   64.46%           
> ================================================
>   Files                    ?        9           
>   Lines                    ?      515           
>   Branches                 ?        0           
> ================================================
>   Hits                     ?      332           
>   Misses                   ?      183           
>   Partials                 ?        0
> ```
> 
> [Impacted Files](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97?src=pr&el=tree) 	Coverage Δ 	
> [src/clustering/run_clust.jl](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97/diff?src=pr&el=tree#diff-c3JjL2NsdXN0ZXJpbmcvcnVuX2NsdXN0Lmps) 	`88.54% <100%> (ø)` 	
> [src/utils/load_data.jl](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97/diff?src=pr&el=tree#diff-c3JjL3V0aWxzL2xvYWRfZGF0YS5qbA==) 	`71.42% <16.66%> (ø)` 	
> [src/clustering/extreme_vals.jl](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97/diff?src=pr&el=tree#diff-c3JjL2NsdXN0ZXJpbmcvZXh0cmVtZV92YWxzLmps) 	`84.21% <33.33%> (ø)` 	
> [src/utils/utils.jl](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97/diff?src=pr&el=tree#diff-c3JjL3V0aWxzL3V0aWxzLmps) 	`83.15% <50%> (ø)` 	
> [src/utils/datastructs.jl](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97/diff?src=pr&el=tree#diff-c3JjL3V0aWxzL2RhdGFzdHJ1Y3RzLmps) 	`97.36% <75%> (ø)` 	
> 
> [Continue to review full report at Codecov](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97?src=pr&el=continue).
> 
> > **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> > `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> > Powered by [Codecov](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97?src=pr&el=footer). Last update [d483c9e...c53c37c](https://codecov.io/gh/holgerteichgraeber/ClustForOpt.jl/pull/97?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

This happened because the `documentation` branch does not have coverage, that's to be expected..

File: .travis.yml [hunk 1]
```diff
@@ -9,23 +9,22 @@ julia:
...
       script:
         - julia --project=docs/ -e 'using Pkg; Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate()'
         - julia --project=docs/ docs/make.jl
-      after_success:
-services: docker
+      after_success: skip
+
+after_success:
+    - julia -e 'using Pkg; Pkg.add("Coverage")'
+    - julia -e 'using Coverage; Codecov.submit(process_folder())'
```
Comment (YoungFaithful) [inline .travis.yml]: Nice

File: Project.toml [hunk 2]
```diff
@@ -12,11 +12,21 @@ Clustering = "aaaa29a8-35af-508c-8bc3-b662a17a0fe5"
...
+[extras]
+Cbc = "9961bab8-2fa3-5c5a-9d89-47fab24efd76"
+JLD2 = "033835bb-8acc-5ee8-8aae-3f567f8a3819"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
+StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
+Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
+
+[targets]
+test = ["Pkg", "Test", "StatsBase", "Cbc", "Random", "JLD2"]
```
Comment (YoungFaithful) [inline Project.toml]: Great use

File: src/clustering/extreme_vals.jl [hunk 3]
```diff
@@ -20,7 +20,7 @@ function simple_extr_val_sel(data::ClustData,
   elseif rep_mod_method=="append"
     data_mod = input_data_modification(data,idcs)
   else
-    @error("rep_mod_method - "*rep_mod_method*" - does not exist")
+    error("rep_mod_method - "*rep_mod_method*" - does not exist")
```
Comment (YoungFaithful) [inline src/clustering/extreme_vals.jl]: What is the advantage?

File: test/extreme_vals.jl [hunk 4]
```diff
@@ -0,0 +1,123 @@
+
+ts_input_data = load_timeseries_data(:CEP_GER18)
```
Comment (YoungFaithful) [inline test/extreme_vals.jl]: Nice

File: test/extreme_vals.jl [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            @test all(ts_clust_extr.weights[1:ts_clust_res.clust_data.K] .≈ ts_clust_res.clust_data.weights)
+            @test all(ts_clust_extr.weights[ts_clust_res.clust_data.K+1:ts_clust_res.clust_data.K+3] .≈ extr_vals.weights)
+         end
+
+
+        end
+    end
+
+
+end
```
Comment (YoungFaithful) [inline test/extreme_vals.jl]: Really good testing

File: src/clustering/run_clust.jl [hunk 6]
```diff
@@ -372,14 +372,14 @@ function run_clust_kmedoids_exact_medoid(
     data_norm::ClustDataMerged,
     n_clust::Int,
     iterations::Int;
-    gurobi_opt=0
+    kmexact_optimizer=0
```
Comment (YoungFaithful) [inline src/clustering/run_clust.jl]: Can we not use:
``kmexact_optimizer::DataType=DataType`` or something else defining the Type?

File: test/test_utils.jl [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    @test t.T == ref.T
+    @test all(t.weights .≈ ref.weights)
+    @test all(t.delta_t .≈ ref.delta_t)
+    @test all(t.k_ids .≈ ref.k_ids)
+    for (k,v) in t.data
+        @test all(t.data[k] .≈ ref.data[k])
+        @test all(t.mean[k] .≈ ref.mean[k])
+        @test all(t.sdv[k] .≈ ref.sdv[k])
+    end
+end
```
Comment (YoungFaithful) [inline test/test_utils.jl]: That's great

File: test/load_data.jl [hunk 8]
```diff
@@ -0,0 +1,21 @@
+
+@testset "load_timeseries_data" begin
+    # load one and check if all the fields are correct
+    ts_input_data = load_timeseries_data(:CEP_GER1)
+    @test ts_input_data.K == 366
+    @test ts_input_data.T == 24
+    @test all(ts_input_data.weights .== 1.)
+
+    # load all four by name - just let them run and see if they run without error
+    ts_input_data = load_timeseries_data(:DAM_GER)
```
Comment (YoungFaithful) [inline test/load_data.jl]: I like the usage of a Symbol there. I think I will do the same in CapExp

File: test/run_clust.jl [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+# make sure to put Random.seed!() before every evaluation of run_clust, because most of them use random number generators in clustering.jl
+# Random.seed!() should give the same result even on different machines:https://discourse.julialang.org/t/is-rand-guaranteed-to-give-the-same-sequence-of-random-numbers-in-every-machine-with-the-same-seed/12344
+
+# put DAM_GER, DAM_CA, CEP GER1, CEP GER18 all years into data.jl file that just tests kmeans with all data inputs
+
+
+@load normpath(joinpath(dirname(@__FILE__),"reference_generation","run_clust.jld2")) reference_results
+
+Random.seed!(1111)
+@testset "run_clust $data" for data in ["CEP_GER1","CEP_GER18"] begin
```
Comment (YoungFaithful) [inline test/run_clust.jl]: you could just use ``data in [:CEP_GER1, :CEP_GER18]``

File: Project.toml (hunk 2 above)
Comment (holgerteichgraeber) [reply Project.toml]: So apparently if you use Pkg.test, you need [extras] and [targets], and if you use `julia runtests.jl`, you need a Project.toml file in the test folder. Somehow Pkg.test fails the tests for us whereas the second option works for all tests (no idea why), I'll open an issue with example in julia discourse on that...

File: src/clustering/extreme_vals.jl (hunk 3 above)
Comment (holgerteichgraeber) [reply src/clustering/extreme_vals.jl]: `@error` is for logging purposes (debugging I guess) and does not throw an exception, whereas `error(msg)` throws an exception. try catch statements work with exceptions. If you want to specify the error type, need to use throw(). For us, I think ErrorException as thrown by error() as the type is ok.

File: src/clustering/run_clust.jl (hunk 6 above)
Comment (holgerteichgraeber) [reply src/clustering/run_clust.jl]: Not sure what you are trying to point to. 
The issue we had was that the optimizer object itself does not have a specific type..

File: test/run_clust.jl (hunk 9 above)
Comment (holgerteichgraeber) [reply test/run_clust.jl]: Good idea

File: test/load_data.jl (hunk 8 above)
Comment (holgerteichgraeber) [reply test/load_data.jl]: Sounds good. I liked the idea of having a different type for given input, having two strings could be confusing to the user.

**Comment by holgerteichgraeber** [comment]
> 
> 
> Great testing overall! I added a few questions and suggestions to the code.
> What would you think about a few comments explaining the main process that you use for testing?

Thanks. The comments for testing mostly in `test/run_clust.jl` or where would you think they are most appropriate?

**Comment by holgerteichgraeber** [comment]
Added some minor updates on testing and documentation. Already pushed to dev.

File: src/clustering/run_clust.jl (hunk 6 above)
Comment (YoungFaithful) [reply src/clustering/run_clust.jl]: I think that changed with the new JuMP update (https://github.com/YoungFaithful/CapacityExpansion.jl/blob/d23ffab3f915d6f02e128774b6d2ae82d1c358f0/src/optim_problems/run_opt.jl#L86)

File: test/reference_generation/run_clust.jl [hunk 10]
```diff
@@ -6,8 +6,8 @@ using Random
 reference_results = Dict{String,Any}()
 
 Random.seed!(1111)
-for data in ["CEP_GER1","CEP_GER18"]
-    ts_input_data = load_timeseries_data(Symbol(data))
+for data in [:CEP_GER1,:CEP_GER18]
```
Comment (YoungFaithful) [inline test/reference_generation/run_clust.jl]: :+1:

**Comment by YoungFaithful** [comment]
> > Great testing overall! I added a few questions and suggestions to the code.
> > What would you think about a few comments explaining the main process that you use for testing?
> 
> Thanks. The comments for testing mostly in `test/run_clust.jl` or where would you think they are most appropriate?

I thought about an explanation in `test/runtests.jl` In case someone else wants to expand the package with their own method and needs to know, how testing works in general.

File: src/clustering/run_clust.jl (hunk 6 above)
Comment (holgerteichgraeber) [reply src/clustering/run_clust.jl]: That's great. Once we have done the renaming of the package, let's make that change. Opening an issue #111

**Comment by holgerteichgraeber** [comment]
> 
> 
> > > Great testing overall! I added a few questions and suggestions to the code.
> > > What would you think about a few comments explaining the main process that you use for testing?
> > 
> > 
> > Thanks. The comments for testing mostly in `test/run_clust.jl` or where would you think they are most appropriate?
> 
> I thought about an explanation in `test/runtests.jl` In case someone else wants to expand the package with their own method and needs to know, how testing works in general.

Sounds good. Opening an issue to be addressed after name change. #112

File: Project.toml (hunk 2 above)
Comment (holgerteichgraeber) [reply Project.toml]: Btw, I just scrolled through the documentation, it seems that this will be streamlined in julia 1.2

**Comment by holgerteichgraeber** [comment]
Thank you for your comments on this @YoungFaithful . Because this PR was previously merged #98 , closing it here.

**Pull request closed without merging**